### PR TITLE
Rule action appliance

### PR DESCRIPTION
**Pull request by stonebuzz**
Add capability to aassign appliance from rule (on associated element)
Action : assign,  regex

image

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | internal 21617

### CODE REVIEW

File: inc/ruleticket.class.php [hunk 1]
```diff
@@ -357,6 +362,28 @@ function executeActions($output, $params, array $input = []) {
...
+                                                                     $this->regex_results[0]);
+                     } else {
+                        $regexvalue = $action->fields["value"];
+                     }
+
+                     if (!is_null($regexvalue)) {
+                        $appliances = new Appliance();
+                        $target_appliances = $appliances->find(["name" => $regexvalue, "is_helpdesk_visible" => true]);
+
+                        if (count($target_appliances) > 0) {
```
[the remark this answers is not in the document]
Comment (stonebuzz) [reply inc/ruleticket.class.php]: done

File: inc/ruleticket.class.php [hunk 2]
```diff
@@ -238,6 +238,12 @@ function executeActions($output, $params, array $input = []) {
                      }
                   }
 
+                  // special case of appliance
+                  if ($action->fields["field"] == "assign_appliance") {
+                     $output["items_id"] = [];
```
[the remark this answers is not in the document]
Comment (stonebuzz) [reply inc/ruleticket.class.php]: done

File: inc/ruleticket.class.php [hunk 3]
```diff
@@ -255,7 +257,17 @@ function executeActions($output, $params, array $input = []) {
...
                             = $action->fields["value"];
                   }
-                  $output[$actions[$action->fields["field"]]["appendto"]][] = $value;
+
+                  // special case of appliance
+                  if ($action->fields["field"] === "assign_appliance") {
+                     if (!array_key_exists("items_id", $output)
+                        || (array_key_exists("items_id", $output) && !is_array($output['items_id']))) {
+                        $output["items_id"] = [];
+                     }
```
Comment (cedric-anne) [inline inc/ruleticket.class.php:266]: If `$output['items_id']` has an invalid type, it is not a normal situation. Calling `trigger_error` will add a trace in GLPI log, which could help to fix this problem.
```suggestion
                     if (!array_key_exists("items_id", $output) || !is_array($output['items_id'])) {
                        if (!is_array($output['items_id'])) {
                           trigger_error('Invalid "items_id" output variable found in "assign_appliance" rule.', E_USER_WARNING);
                        }
                        $output["items_id"] = [];
                     }
```

Comment (stonebuzz) [reply inc/ruleticket.class.php]: done

File: inc/ruleticket.class.php [hunk 4]
```diff
@@ -704,6 +741,14 @@ function getActions() {
       $actions['affectobject']['force_actions']             = ['affectbyip', 'affectbyfqdn',
                                                                     'affectbymac'];
 
+      $actions['assign_appliance']['name']                  = _n('Associated element', 'Associated elements', Session::getPluralNumber())." : ".Appliance::getTypeName(1);
```
Comment (cedric-anne) [inline inc/ruleticket.class.php:744]:
```suggestion
      $actions['assign_appliance']['name']                  = __('Associated appliances');
```

**Review by cedric-anne — APPROVED** [review]

**Review by orthagh — APPROVED** [review]

**Comment by trasher** [comment]
@stonebuzz could you please rebase this one to solve the conflict? thanks!

**Comment by stonebuzz** [comment]
> @stonebuzz could you please rebase this one to solve the conflict? thanks!

 of course

**Review by trasher — APPROVED** [review]

**Pull request merged**